File: src/minymph_project/Moves.java
package minymph_project;

import java.util.List;
import java.util.ArrayList;
import java.util.function.Consumer;
import javax.swing.*;
import java.util.Random;

/**
 * Represents a move in the Minymph game with attributes such as name, damage, PP (Power Points),
 * critical rate, and any potential side effect.
 * Moves can be chosen by players and listed for reference.
 */
public class Moves {
    
    private static List<Moves> normalMoves = new ArrayList<>();
    private static List<Moves> aiNormalMoves = new ArrayList<>();
    private GUIBattle gui;
    private String name;
    private int damage;
    private int PP;
    private float critRate;
    private String sideEffect;
    private float accuracy;
    private Player owner;

    String moveChoice = 
            "1 : Headbutt \n" +
            "2 : Sweep \n" +
            "3 : Haze \n" +
            "4 : HyperRush \n" +
            "-1 : Go back \n";

    /**
     * Constructs a new Move with specified attributes.
     *
     * @param gui the GUI associated with the game
     * @param name the name of the move
     * @param damage the damage value of the move
     * @param PP the Power Points of the move
     * @param critRate the critical rate of the move
     * @param sideEffect any side effect the move may have
     * @param accuracy accuracy of the move
     * @param owner owner of the minymph
     */
    public Moves(GUIBattle gui, String name, Integer damage, Integer PP, Float critRate, Float accuracy, String sideEffect, Player owner) {
        this.gui = gui;
        this.name = name;
        this.damage = damage;
        this.PP = PP;
        this.critRate = critRate;
        this.sideEffect = sideEffect;
        this.accuracy = accuracy;
        this.owner = owner;
    }

    /**
     * Constructs a new Move associated with a GUI.
     *
     * @param gui the GUI associated with the game
     */
    public Moves(GUIBattle gui) {
        this.gui = gui;
    }

    /**
     * Gives the move's minymph owner (to avoid sharing move's pp)
     * @return owner of a minymph's move
     */
    public Player getOwner()
    {
    	return this.owner;
    }
    /**
     * Default constructor for creating a Move.
     */
    public Moves() {
    }

    /**
     * Returns the name of the move.
     *
     * @return the name of the move
     */
    public String getName() {
        return this.name;
    }

    /**
     * Returns the damage value of the move.
     *
     * @return the damage value of the move
     */
    public int getDamage() {
        return this.damage;
    }

    /**
     * Returns the Power Points (PP) of the move.
     *
     * @return the PP of the move
     */
    public int getPP() {
        return this.PP;
    }
    
    /**
     * Set PP of a move
     * @param PP value to be entered
     */
    public void setPP(Integer PP)
    {
    	if (this.PP<0 && this.getOwner() instanceof ConcretePlayer)
    	{
    		this.PP = 0; // make sure the PP value is always >=0
    	}
    	this.PP = PP;
    }
    
    /**
     * get PP of ai's minymphs
     * @return pp of a specified ai's minymph
     */
    public int getPPai()
    {
    	if (this.getOwner() instanceof AI)
    	{
    		return this.PP;
    	}
    	return 0;
    }
    
    /**
     * Set PP of a move for ai (avoid sharing move's pp!)
     * @param PP value to be entered
     */
    public void setPPai(Integer PP)
    {
    	if (this.PP<0)
    	{
    		this.PP = 0; // make sure the PP value is always >=0
    	}
    	if (this.getOwner() instanceof AI)
    	{
    		this.PP = PP;
    	}
    }
    
    /**
     * Returns the accuracy of the move.
     *
     * @return the accuracy of the move
     */
    public float getAccuracy() {
        return this.accuracy;
    }

    /**
     * Returns the critical rate of the move.
     *
     * @return the critical rate of the move
     */
    public float getCritRate() {
        return this.critRate;
    }

    /**
     * 
     * @param moveAccuracy the selected move accuracy 
     * @return true if the move is executed based on its accuracy
     */
	public static Boolean accuracyCheck(Float moveAccuracy)
	{
		Random random = new Random();
		int randomNumber = random.nextInt(101);
		float floatDebug = (float)(randomNumber);
		if (floatDebug<=moveAccuracy)
		{
			return true;
		}
		return false;
		
	}
	
    
    /**
     * Returns the side effect of the move, if any.
     *
     * @return the side effect of the move
     */
    public String getSideEffect() {
        return this.sideEffect;
    }

    /**
     * Prints the name of the move to the console.
     */
    public void printMove() {
        System.out.println(this.name);
    }

    /**
     * Prints the list of all available normal type moves to the console.
     */
    public static void printList() {
        System.out.println("Available normal type moves:");
        for (Moves mov : normalMoves) {
            System.out.println(mov.name);
        }
    }

    /**
     * Adds this move to the player's move list
     */
    public void addToList() {
        normalMoves.add(this);
    }
    
    /**
     * Adds this move to the AI move's list.
     */
    public void addToAIList() {
        aiNormalMoves.add(this);
    }
    
    /**
     * get player's minymph moves
     * @return the player's minymph moves
     */
    public static List<Moves> getMoves()
    {
    	return normalMoves;
    }
    
    /**
     * get ai minymph moves
     * @return the ai minymph moves
     */
    public static List<Moves> getAIMoves()
    {
    	return aiNormalMoves;
    }

    /**
     * Opens a dialog for the user to choose a move from the list of normal moves.
     * Once selected, the move is passed to a callback function.
     *
     * @param callback the function to process the selected move
     */
    public void chooseMove(Consumer<Moves> callback) {
        JDialog moveDialog = new JDialog(gui, "Choose a Move", true);
        moveDialog.setSize(300, 200);
        moveDialog.setLayout(new BoxLayout(moveDialog.getContentPane(), BoxLayout.Y_AXIS));

        for (Moves move : normalMoves) {
            JButton moveButton = new JButton(move.getName());
            moveButton.addActionListener(e -> {
                callback.accept(move);
                moveDialog.dispose();
            });
            moveDialog.add(moveButton);
        }

        moveDialog.setVisible(true);
    }
}

File: core/Moves.py
import random
from typing import TYPE_CHECKING, List, Optional, Callable
import tkinter as tk
from tkinter import ttk

if TYPE_CHECKING:
    from ..game.GuiBattle import GuiBattle
    from ..characters.Player import Player
    from ..characters.ConcretePlayer import ConcretePlayer

class Moves:
    """
    Represents a move in the Minymph game with attributes such as name, damage, PP (Power Points),
    critical rate, and any potential side effect.
    Moves can be chosen by players and listed for reference.
    """

    # Attributs statiques (partagés par toutes les instances)
    _normal_moves: List['Moves'] = []  # Liste des mouvements normaux (joueur)
    _ai_normal_moves: List['Moves'] = []  # Liste des mouvements normaux (IA)

    # Texte pour le choix des mouvements
    MOVE_CHOICE: str = (
        "1 : Headbutt \n"
        "2 : Sweep \n"
        "3 : Haze \n"
        "4 : HyperRush \n"
        "-1 : Go back \n"
    )

    def __init__(
        self,
        gui: Optional['GuiBattle'] = None,
        name: str = "",
        damage: int = 0,
        pp: int = 0,
        crit_rate: float = 0.0,
        accuracy: float = 0.0,
        side_effect: str = "",
        owner: Optional['Player'] = None
    ) -> None:
        """
        Constructs a new Move with specified attributes.

        Args:
            gui: The GUI associated with the game (optional).
            name: The name of the move (optional).
            damage: The damage value of the move (optional).
            pp: The Power Points of the move (optional).
            crit_rate: The critical rate of the move (optional).
            accuracy: The accuracy of the move (optional).
            side_effect: Any side effect the move may have (optional).
            owner: The owner of the move (optional).
        """
        self.gui: Optional['GuiBattle'] = gui
        self.name: str = name
        self.damage: int = damage
        self.pp: int = pp
        self.crit_rate: float = crit_rate
        self.accuracy: float = accuracy
        self.side_effect: str = side_effect
        self.owner: Optional['Player'] = owner

    @classmethod
    def get_normal_moves(cls) -> List['Moves']:
        """
        Returns the list of all normal moves (player).

        Returns:
            The list of normal moves.
        """
        return cls._normal_moves

    @classmethod
    def get_ai_moves(cls) -> List['Moves']:
        """
        Returns the list of all AI moves.

        Returns:
            The list of AI moves.
        """
        return cls._ai_normal_moves

    def get_owner(self) -> Optional['Player']:
        """
        Returns the owner of this move.

        Returns:
            The owner Player object.
        """
        return self.owner

    def get_name(self) -> str:
        """
        Returns the name of the move.

        Returns:
            The name of the move.
        """
        return self.name

    def get_damage(self) -> int:
        """
        Returns the damage value of the move.

        Returns:
            The damage value.
        """
        return self.damage

    def get_pp(self) -> int:
        """
        Returns the Power Points (PP) of the move.

        Returns:
            The PP value.
        """
        return self.pp

    def set_pp(self, pp: int) -> None:
        """
        Sets the Power Points (PP) of the move.

        Args:
            pp: The new PP value.
        """
        if self.pp < 0 and isinstance(self.owner, ConcretePlayer):
            self.pp = 0  # Ensure PP is always >= 0 for player's moves
        self.pp = pp

    def get_pp_ai(self) -> int:
        """
        Returns the Power Points (PP) of the move for AI.

        Returns:
            The PP value if the owner is AI, otherwise 0.
        """
        if isinstance(self.owner, Ai):
            return self.pp
        return 0

    def set_pp_ai(self, pp: int) -> None:
        """
        Sets the Power Points (PP) of the move for AI.

        Args:
            pp: The new PP value.
        """
        if pp < 0:
            self.pp = 0  # Ensure PP is always >= 0
        if isinstance(self.owner, Ai):
            self.pp = pp

    def get_accuracy(self) -> float:
        """
        Returns the accuracy of the move.

        Returns:
            The accuracy value.
        """
        return self.accuracy

    def get_crit_rate(self) -> float:
        """
        Returns the critical rate of the move.

        Returns:
            The critical rate value.
        """
        return self.crit_rate

    @staticmethod
    def accuracy_check(move_accuracy: float) -> bool:
        """
        Checks if a move hits based on its accuracy.

        Args:
            move_accuracy: The accuracy of the move.

        Returns:
            True if the move hits, False otherwise.
        """
        random_number = random.randint(0, 100)
        return random_number <= move_accuracy

    def get_side_effect(self) -> str:
        """
        Returns the side effect of the move.

        Returns:
            The side effect.
        """
        return self.side_effect

    def print_move(self) -> None:
        """
        Prints the name of the move to the console.
        """
        print(self.name)

    @staticmethod
    def print_list() -> None:
        """
        Prints the list of all available normal type moves to the console.
        """
        print("Available normal type moves:")
        for move in Moves._normal_moves:
            print(move.name)

    def add_to_list(self) -> None:
        """
        Adds this move to the player's move list.
        """
        Moves._normal_moves.append(self)

    def add_to_ai_list(self) -> None:
        """
        Adds this move to the AI's move list.
        """
        Moves._ai_normal_moves.append(self)

    def choose_move(self, callback: Callable[['Moves'], None]) -> None:
        """
        Opens a dialog for the user to choose a move from the list of normal moves.
        Once selected, the move is passed to a callback function.

        Args:
            callback: The function to process the selected move.
        """
        if not self.gui:
            return

        # Créer une fenêtre de dialogue
        dialog = tk.Toplevel(self.gui)
        dialog.title("Choose a Move")
        dialog.geometry("300x200")
        dialog.transient(self.gui)  # Empêcher l'interaction avec la fenêtre parente
        dialog.grab_set()  # Capturer les événements clavier/souris

        # Disposition verticale
        frame = tk.Frame(dialog)
        frame.pack(fill=tk.BOTH, expand=True)

        for move in Moves._normal_moves:
            button = tk.Button(
                frame,
                text=move.get_name(),
                command=lambda m=move: self._on_move_selected(m, callback, dialog)
            )
            button.pack(fill=tk.X, pady=2)

    def _on_move_selected(self, move: 'Moves', callback: Callable[['Moves'], None], dialog: tk.Toplevel) -> None:
        """
        Callback pour la sélection d'un mouvement.

        Args:
            move: Le mouvement sélectionné.
            callback: Fonction à appeler avec le mouvement sélectionné.
            dialog: La fenêtre de dialogue à fermer.
        """
        callback(move)
        dialog.destroy()